### ISSUE TITLE
Boost performance for API apps (CakeResponse in route)

### ISSUE THREAD

**Issue by josephzidell** [cakephp/cakephp#10833]
This is a (multiple allowed):

* [ ] bug
* [x] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 2.10.0

### What you did
Use a closure in a route to directly render a CakeResponse. Gives huge performance boost and more control, esp. in API environments.

```php
// routes.php
Router::connect('/api/v1/orders/:id/', ['callback' => function ($id) {
	return new CakeResponse(['body' => \MyApp\Api\V1\Order::get($id)]);
}, '[method]' => 'GET'], ['pass' => ['id']]);
```

This is a simplified example, and userland wrapper can simplify further:
```php
// routes.php
Route::get('/api/v1/orders/:id/', function ($id) {
	return new CakeResponse(['body' => \MyApp\Api\V1\Order::get($id)]);
});
```
I was planning on opening a PR for 3.x, but got to 2.x first since it's closing soon.

PHP 5.3 friendly PR is adding 1 new line with an `if` condition:
```php
// Dispatcher.php:158
if (isset($request->params['callback'])) {
	$response = call_user_func_array($request->params['callback'], $request->params['pass']);
} else {
	$controller = $this->_getController($request, $response);
	...
	$response = $this->_invoke($controller, $request);
}
```

If you guys see the value in this, I can open PRs for 2.x and 3.x

**Comment by josephzidell** [comment]
Laravel: https://laravel.com/docs/5.4/routing#basic-routing
FuelPHP: https://fuelphp.com/docs/general/routing.html#/inline_routes
Symfony: https://symfony.com/doc/current/components/routing.html#routes-as-closures

**Comment by markstory** [comment]
I am not sure I would call the option 'callback'. Isn't the callback a very stripped down 'controller'? Similar discussions have been happening for 3.x already (#10689 & #10643).

What would be the motivation to try and get this added to 2.x now? Wouldn't it be simpler to focus on the releases that are going to be more actively used in the future?

**Comment by josephzidell** [comment]
I can call it `'closure'`. We're building an API in a 2.x app and made this change. It's a really small change with a lot of value.

**Comment by ADmad** [comment]
In 3.x you can use the middleware to return response and 3.5 will have support for routing based middlewares which would allow achieving the same as requested. I would add this feature to router itself only if it provides significant improvement over using middleware (which I doubt).

**Comment by ADmad** [comment]
:-1: to adding new features to 2.x.

**Comment by josephzidell** [comment]
@ADmad I understand that middleware makes sense for 3.x. Why the 👎 for 2.x?

**Comment by josephzidell** [comment]
It's such a basic feature that every other framework has. We're trying to migrate from 1.x to 2.x right now (400 models, 400 controllers, 1500 views) and I'm trying to help with small things that add value.

**Comment by ADmad** [comment]
We had kept developing (adding features) 2.x long enough, can't keep doing that forever. 2.10 is the last feature releases for 2.x and 2.10 RC has already been released.

Why can't you just start your new API with 3.x? Incrementally implementing parts of the app with latest version is how upgrading large apps should be handled.

**Comment by josephzidell** [comment]
The API needs to share business logic with 2.x, so a separate app won't work. It needs to be folded into the rest of the app.

I'm very much in favor of extending functionality in userland, but the namespaceless nature of 2.x makes it impossible without modifying core files. I'm trying to have an open line of communication here so that real world value can get added even later down the road.

I understand not wanted to maintain 2.x. I have plenty experience maintaining software and I hate having to deal with bugs on old. But when someone comes with a PR that will help him and others, I literally have no reason to turn it down.

You guys work super hard for OSS, which deserves all medals. Your openness to accepting others' contributions will go a long way to making CakePHP the best that it can be.

**Comment by markstory** [comment]
@josephzidell I know my hesitation in adding new 'large' features to 2.x comes down to long term maintenance and the potential to destabilize existing features. 

Given that routeable middleware works quite a bit differently in 3.x I have concerns about adding a feature that doesn't have a clean upgrade path, which is where my reservations come from.

**Comment by josephzidell** [comment]
@markstory & @ADmad I completely understand.

Can we work together to design it in an easily upgradable manner? For example, in 3.x the routes could auto send to middleware, or middleware could read from the routes. Something like that.

My point it, instead of rejecting it, let's see if we can make it work. There are a lot of 2.x folks out there.

**Comment by bham1586** [comment]
I have a very large 2.x application, and have no plans to move to 3.x anytime soon because it will take weeks (if not months) of work. I can see, not wanting to add large features to the 2.x series, but not accepting useful PRs like this seems silly.

**Comment by markstory** [comment]
@bham1586 This isn't a pull request yet. It is an issue discussion. Lets get to a solution that would satisfy the requirement for a reasonable upgrade path to 3.x before we start talking about rejecting pull requests.

@josephzidell I'm open to working on a solution. But I want to be able to deliver on the promise of shipping 2.10 in the next few weeks.

**Comment by lorenzo** [comment]
> I'm very much in favor of extending functionality in userland, but the namespaceless nature of 2.x makes it impossible without modifying core files.

@josephzidell how does having closures in the router help with the namespaces issue?

As a side note, I have implemented in the past a feature for route that execute any callable, all in user land and without modifying the core. it does not take more than a couple hours to put together

**Comment by josephzidell** [comment]
@lorenzo Because I cannot override `Dispatcher.php`

**Comment by josephzidell** [comment]
@markstory Seems there are 2 paths: (a) 2.x -> 3.x and (b) 3.x -> 3.5

Does the `routes.php` file need to look the same across both upgrade paths or do you just need 3 PRs?

**Comment by ADmad** [comment]
> We're trying to migrate from 1.x to 2.x right now...

Just to add some perceptive to that, 2.0.0 was released on October 16 2011 and 3.0.0 on Mar 22 2015.

**Comment by lorenzo** [comment]
@josephzidell I implemented it using dispatcher filters. All the dispatching process can be bypassed using dispatcher filters. That is how the `AssestDispatcher` works, for instance.

**Comment by josephzidell** [comment]
@ADmad I know... Out enterprise application is large and it took 3 years of convincing to even get to 2.x. The reason we can't go to 3.x is due to arrays vs objects, even though I much prefer objects. If not for that, we'd be heading to 3.x land.

**Comment by josegonzalez** [comment]
I did this with a custom dispatch filter in CakePHP 2: http://josediazgonzalez.com/2014/03/23/cakephp-a-microframework/

I think for 3.x would be a good core addition - speed up your apps with this one neat trick! - but in 2.x, might be too large to add to core. I'll leave that decision @lorenzo @ADmad and @markstory though.

**Comment by josephzidell** [comment]
@josegonzalez I'll look at your dispatch filter. I do like the sleezy ad pitch though 😆

**Comment by josephzidell** [comment]
@josegonzalez Looked though your code. Should have been added to the core years ago.

**Comment by markstory** [comment]
Given that the desired features can be built as a plugin or application code is there a burning need to have a similar feature in CakePHP?

**Comment by josephzidell** [comment]
I'll try it out tomorrow

**Comment by bancer** [comment]
> Why can't you just start your new API with 3.x? Incrementally implementing parts of the app with latest version is how upgrading large apps should be handled.

Due to very complex business logic that is used in different parts of our application I also do not see a clear path of migrating our code to 3.x. That would take months if not years. With such high number of commits that we have every day it is not possible to stall any changes even for a week. 200 models, 150 controllers, 50 components. Maintaining two different versions of the application is not possible from the business point of view as the same changes would need to be done in 2.x and 3.x versions for the time of the migration. I think we will never be able to migrate :(. 

The only possible way would be to make possible to run 2.x code in 3.x with some sort of adapter. At least for the period of migration.

**Comment by josephzidell** [comment]
I agree. Without an ORM adapter, cakephp 3 for us is very far off indeed.

**Comment by josephzidell** [comment]
Logistically, we'd need an opt-in adapter that will return either arrays or objects that we can specify. Same thing for saving. Needs to support both arrays and objects

**Comment by josegonzalez** [comment]
@bancer @josephzidell Unfortunately, providing such a shim for Cake3 to produce arrays in the structure you'd expect from Cake2 would be difficult, and I personally would prefer that the core developers continue working on Cake3 related bugs and enhancements (though would welcome such a change from the community).

A few notes around upgrading:

- You can incrementally upgrade your application by rewriting pieces at a time in the new framework. Either new features can be done in a separate application, or you can pull in the new libraries to rewrite old code. IIRC thats how @lorenzo and @jippi rewrote a _fairly_ large CakePHP application.
- I would move db query responsibility away from business logic. Using custom finders is a good way to get rid of most of that logic within your application code, as then you're hiding the vagaries of each specific lookup within it's own method. You could take this a step further and move them into traits that you use in your Cake2 model classes, and have those traits also implement the Cake3 find as well. The naming in Cake2 has an underscore, while Cake3 does not, so you can have them side by side in the same class.
- Refactor your business logic from CakePHP. I see that lots of times people have hundred/two hundred line controller actions that are hard to read, understand, and modify. Most of it with embedded CakePHP-related code, when all of that stuff can be injected. Moving that into classes around what the logic actually does is a good way to simplify the codebase, add code coverage, and reduce the dependency around any single CakePHP version.

I've done a similar migration for a large (300 model) CakePHP 1.3 application (to 3.x) and though it did take a month or two, it was possible to do side-by-side with the existing application. I did have the luxury of good tests, and we took some time up front to refactor yucky code, but it was possible (team of three, and I was doing this part-time).

Semi-related, I used to work on [cakephp-entity](https://github.com/josegonzalez/cakephp-entity) for 2.x, but I'm sure it has issues and haven't dedicated any time to it since I made the jump. If you think such a thing would help your team adopt the newer-style ORM, would be great to have some input.

Regarding the original issue, I think maybe we've gone off topic. @markstory I'm assuming this won't make it to the 2.10.x release, so should we close the issue or leave it open for other discussion?

**Comment by markstory** [comment]
I don't think a significant rebuild of routing is going to land before 2.10 is released. We can leave this open until then as it has created some great discussion.

**Comment by markstory** [comment]
Closing as 2.10 was the last feature release for 2.x and this wasn't ready for that release.
